### PR TITLE
Freindface fullstack

### PR DESCRIPTION
**Pull request by Danielcoates-scottlogic**

### CODE REVIEW

File: friendface-web/src/main/java/com/example/friendface/Post.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Column
+    private Date date = new Date();
+
+    @ManyToOne
+    @JoinColumn(name = "username", referencedColumnName = "username", nullable = false)
+    private User author;
+
+    public Date getDate() {
+        return date;
```
Comment (Danielcoates-scottlogic) [inline friendface-web/src/main/java/com/example/friendface/Post.java]: g

File: friendface-web/src/main/java/com/example/friendface/Post.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package com.example.friendface;
+
+import jakarta.persistence.*;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/Post.java:3]: Try to avoid using wildcard imports (.*) - if IntelliJ does these automatically you can configure the settings to prevent this.

File: friendface-web/src/main/java/com/example/friendface/PostController.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package com.example.friendface;
+
+
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/PostController.java]: Wildcard import

File: friendface-web/src/main/java/com/example/friendface/PostController.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.List;
+@CrossOrigin(origins = "http://localhost:4200")
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/PostController.java:15]: This is correct, but this is the only one you need if the origin doesn't change. If you define this on the class level, you shouldn't need to redefine it

File: friendface-web/src/main/java/com/example/friendface/PostController.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    }
+    @CrossOrigin(origins = "http://localhost:4200")
+    @GetMapping
+    public ResponseEntity<List<Post>> getPosts() {
+        List<Post> posts = this.postService.getPosts();
+        return new ResponseEntity<>(posts,HttpStatus.OK);
+    }
+    @CrossOrigin(origins = "http://localhost:4200")
+    @PostMapping
+    public ResponseEntity<Post> addPost(@RequestBody Post post) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/PostController.java]: Could add some validation here using `@Valid`

File: friendface-web/src/main/java/com/example/friendface/PostService.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    private final PostRepository postRepository;
+
+    public PostService(PostRepository postRepository) {
+        this.postRepository = postRepository;
+    }
+
+    public List<Post> getPosts(){
+        return postRepository.findAll();
+    }
+    public Post addPost(Post post) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/PostService.java]: Consider null or validation checks on addPost

File: friendface-web/src/main/java/com/example/friendface/SecurityConfig.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+public class SecurityConfig {
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder(12); // 12 is the cost factor
+    }
+    @Bean
+    public UserDetailsService userDetailsService(PasswordEncoder encoder) {
+        UserDetails user = org.springframework.security.core.userdetails.User.builder()
+                .username("swagger")
+                .password(encoder.encode("swagger123"))  // BCrypt-hashed password
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/SecurityConfig.java]: Do not hardcode passwords into your code - use environment variable or config server

File: friendface-web/src/main/java/com/example/friendface/UserController.java [hunk 8]
```diff
@@ -1,24 +1,48 @@
...
+        dto.setUsers(dtoUsers);
+        return new ResponseEntity<>(dto,HttpStatus.OK);
     }
-
-    @PostMapping ResponseEntity<User> addUser(@RequestParam @NotBlank String username) {
-        return this.userService.addUser(username);
+    @CrossOrigin(origins = "http://localhost:4200")
+    @PostMapping
+    public ResponseEntity<User> addUser(@RequestBody CreateUserDto dto) {
+        if (dto.getUsername() == null || dto.getUsername().isEmpty()) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/UserController.java]: This is correct validation, but you can use `spring-boot-starter-validation` which I believe you already use in other areas of your code.

File: friendface-web/src/main/java/com/example/friendface/UserService.java [hunk 9]
```diff
@@ -1,29 +1,30 @@
...
     }
 
-    public ResponseEntity<User> addUser(@RequestParam @NotBlank String username) {
-        User user = new User();
-        user.setUsername(username);
-        userRepository.save(user);
-        return new ResponseEntity<>(user, HttpStatus.OK);
+    public User addUser(User user) {
+        if (user.getUsername().isEmpty()) {
+            return null;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/UserService.java]: Avoid returning `null`. Throw an `IllegalArgumentException` or use `Optional`

File: friendface-web/src/test/java/com/example/friendface/PostControllerTest.java
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/PostControllerTest.java:1]: Good crack at these tests, but currently they only check for the status of posts ('ok' or 'created'), they don't validate the content in any way

File: friendface-web/src/main/java/com/example/friendface/UserController.java [hunk 10]
```diff
@@ -1,24 +1,48 @@
...
+            conversion.setUsername(user.getUsername());
+            dtoUsers.add(conversion);
+        }
+        dto.setUsers(dtoUsers);
+        return new ResponseEntity<>(dto,HttpStatus.OK);
     }
-
-    @PostMapping ResponseEntity<User> addUser(@RequestParam @NotBlank String username) {
-        return this.userService.addUser(username);
+    @CrossOrigin(origins = "http://localhost:4200")
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/UserController.java]: Similar to other controller - if you have the `@CrossOrigin` annotation on the top level, you shouldn't need it here unless the `origins` argument is different

File: friendface-web/src/test/java/com/example/friendface/PostControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        when(postService.getPosts()).thenReturn(List.of(post1, post2));
+
+        ResultActions response = mockMvc.perform(get("/posts"))
+                .andExpect(status().isOk());
+        verify(postService).getPosts();
+    }
+
+
+    @Test
+    public void addPost() throws Exception {
```
Comment (Danielcoates-scottlogic) [inline friendface-web/src/test/java/com/example/friendface/PostControllerTest.java:54]: r

File: friendface-web/src/main/java/com/example/friendface/CreateUserDto.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    @NotEmpty(message = "Password required")
+    private String password;
+
+
+
+    public String getPassword() {
+        return password;
+    }
+
+    public void setPassword(String password) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/CreateUserDto.java]: You can add size constraints here if you want

File: friendface-web/src/main/java/com/example/friendface/User.java [hunk 13]
```diff
@@ -1,13 +1,28 @@
...
 @Entity
 @Table(name="users")
 public class User {
     @Id
     @Column
     private String username;
+    @Column
+    private String colour;
+    @Column
+    private String password;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/User.java:26]: Consider using the `@JsonIgnore` annotation to prevent the password from being returned in api responses

File: friendface-web/src/test/java/com/example/friendface/PostControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        mockMvc.perform(post("/posts")
+                        .with(csrf())
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content("{\"author\":\"Dan\", \"content\":\"post here\"}"))
+                .andExpect(status().isCreated())
+                .andExpect(jsonPath("$.id").value(4))
+                .andExpect(jsonPath("$.user.username").value("dan"))
+                .andExpect(jsonPath("$.user.password").value(""));
+    }
+
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/PostControllerTest.java:109]: Could you potentially add a test that caters for missing fields?

**Comment by MMak1599** [comment]
As an extension, consider looking into Lombok to simplify some of your code, and logging (you can use SLF4J logging)

File: friendface-web/src/main/java/com/example/friendface/Post.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    public String getContents() {
+        return contents;
+    }
+
+    public void setContents(String contents) {
+        this.contents = contents;
+    }
+    public User getUser() {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/Post.java]: Could it be worth renaming this to author, rather than user to keep it a bit more consistent?

File: friendface-web/src/main/java/com/example/friendface/AuthController.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.web.bind.annotation.*;
+@CrossOrigin(origins = "http://localhost:4200")
+@RestController
+@RequestMapping("login")
+public class AuthController {
+
+    @Autowired
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/AuthController.java]: You could probably improve this code with contructor injection, so rather than autowiring all of these individually you could autowire a constructor that takes these in as parameters

File: friendface-web/src/main/java/com/example/friendface/AuthRequest.java [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package com.example.friendface;
+
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class AuthRequest {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/AuthRequest.java:11]: Could potentially look into adding validation annotations for username and password

File: friendface-web/src/main/java/com/example/friendface/AuthResponse.java [hunk 18]
```diff
@@ -0,0 +1,12 @@
+package com.example.friendface;
+
+import lombok.Getter;
+
+@Getter
+public class AuthResponse {
+    private String token;
+
+    public AuthResponse(String token) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/AuthResponse.java]: Good use of constructor - just as an aside, Lombok provides an annotation that can also generate this for you: `@AllArgsConstructor`

File: friendface-web/src/main/java/com/example/friendface/CreateUserDto.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
+package com.example.friendface;
+
+import jakarta.validation.constraints.NotEmpty;
+
+
+public class CreateUserDto {
+
+    @NotEmpty(message = "Username Required")
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/CreateUserDto.java]: `@NotEmpty` is good, but it is worth looking into `@NotBlank` to cover more cases

File: friendface-web/src/main/java/com/example/friendface/CreateUserDto.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    private String username;
+    private String colour;
+    @NotEmpty(message = "Password required")
+    private String password;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/CreateUserDto.java]: Minimum size requirement for password?

File: friendface-web/src/main/java/com/example/friendface/CreateUserDto.java [hunk 21]
```diff
@@ -0,0 +1,39 @@
+package com.example.friendface;
+
+import jakarta.validation.constraints.NotEmpty;
+
+
+public class CreateUserDto {
+
+    @NotEmpty(message = "Username Required")
+    private String username;
+    private String colour;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/CreateUserDto.java]: You could look into initialising these fields (set default values or make this field optional) to avoid NPEs.

File: friendface-web/src/main/java/com/example/friendface/HandleValidationExceptions.java [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+@ControllerAdvice
+public class HandleValidationExceptions {
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    @ResponseBody
+    public Map<String, String> handleException( MethodArgumentNotValidException e) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/HandleValidationExceptions.java:22]: Probably worth logging the exception here. Can use SLF4J

File: friendface-web/src/main/java/com/example/friendface/JwtFilter.java [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        if (username != null && authentication == null) {
+            List<GrantedAuthority> authorities = List.of(new SimpleGrantedAuthority("ADMIN"));
+            User user = loginService.findByUsername(username);
+            if (user != null) {
+                UsernamePasswordAuthenticationToken authToken =
+                        new UsernamePasswordAuthenticationToken(user, null, authorities);
+                SecurityContextHolder.getContext().setAuthentication(authToken);
+                System.out.println("Authorities: " + SecurityContextHolder.getContext().getAuthentication().getAuthorities());
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtFilter.java]: Use logging frameworks, not system out

File: friendface-web/src/main/java/com/example/friendface/JwtFilter.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws ServletException, IOException {
+        final String authHeader = request.getHeader("Authorization");
+
+        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
+            chain.doFilter(request, response);
+            return;
+        }
+        final String token = authHeader.substring(7);
+        final String username = jwtUtil.extractUsername(token);
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtFilter.java]: if there are any issues, this will fail silently. It's worth adding exception handling here - could use a try / catch

File: friendface-web/src/main/java/com/example/friendface/JwtFilter.java [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+        if (username != null && authentication == null) {
+            List<GrantedAuthority> authorities = List.of(new SimpleGrantedAuthority("ADMIN"));
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtFilter.java:53]: Might want to extend this to only grant ADMIN to certain users rather than all

File: friendface-web/src/main/java/com/example/friendface/JwtFilter.java [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+import java.util.List;
+
+@NonNullApi
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtFilter.java]: Do you need this?

File: friendface-web/src/main/java/com/example/friendface/JwtUtil.java [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+        try {
+            byte[] keyBytes = Base64.getDecoder().decode(secret);
+            return Jwts.parserBuilder()
+                    .setSigningKey(keyBytes)
+                    .build()
+                    .parseClaimsJws(token)
+                    .getBody()
+                    .getSubject();
+        } catch (JwtException | IllegalArgumentException e) {
+            throw new IllegalArgumentException("Invalid or expired token");
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtUtil.java:52]: Worth logging at the error level when dealing with token validation (`Logger.error`). Not the most necessary in this case but it's good practice for project work and is incredibly helpful for cases where you may need to debug

File: friendface-web/src/main/java/com/example/friendface/JwtUtil.java [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.util.Date;
+@Component
+public class JwtUtil {
+
+    @Value("${jwt.secret}")
+    private String secret;
+
+    @Value("${jwt.expiration-time}")
+    private long jwtExpiration;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtUtil.java:25]: Worth adding some validation - for example, what would happen if you set this to be a negative number?

File: friendface-web/src/main/java/com/example/friendface/JwtUtil.java [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+
+    public String extractUsername(String token) {
+        try {
+            byte[] keyBytes = Base64.getDecoder().decode(secret);
+            return Jwts.parserBuilder()
+                    .setSigningKey(keyBytes)
+                    .build()
+                    .parseClaimsJws(token)
+                    .getBody()
+                    .getSubject();
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtUtil.java:49]: I mentioned looking into more roles in a previous comment, if you do that then you could also look into the role in the jwt to make it more descriptive and flexible

File: friendface-web/src/main/java/com/example/friendface/JwtUtil.java [hunk 30]
```diff
@@ -0,0 +1,46 @@
...
+@Component
+public class JwtUtil {
+
+    @Value("${jwt.secret}")
+    private String secret;
+
+    @Value("${jwt.expiration-time}")
+    private long jwtExpiration;
+
+    public String generateToken(String username) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/JwtUtil.java:27]: Error handling?

File: friendface-web/src/main/java/com/example/friendface/Like.java [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+import jakarta.persistence.Table;
+import jakarta.validation.constraints.NotEmpty;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Table(name="likes")
+@Getter
+@Setter
+public class Like {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/Like.java]: I can imagine this being fine, but it's worth being more specific with your naming to avoid any crossover / confusion. I think 'Like' is a reserved word in a few different SQL dialects

File: friendface-web/src/main/java/com/example/friendface/Like.java [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column
+    private int id;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/Like.java]: Maybe try to use Integer instead of int. Wrapper classes such as Integer is preferable when it comes to entities to allow for null values, etc.

File: friendface-web/src/main/java/com/example/friendface/Like.java
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/Like.java:1]: As an extension, if you want to track who liked what, and when, you can add a field like createdAt which Spring handles with the @CreatedDate annotation

File: friendface-web/src/main/java/com/example/friendface/LikeDto.java [hunk 33]
```diff
@@ -0,0 +1,11 @@
+package com.example.friendface;
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class LikeDto {
+    private String username;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LikeDto.java:15]: Can add validation here

File: friendface-web/src/main/java/com/example/friendface/LikeDto.java [hunk 34]
```diff
@@ -0,0 +1,11 @@
...
+    private int postId;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LikeDto.java]: I think you define postId as a different type in the Post entity - worth being consistent here.

File: friendface-web/src/main/java/com/example/friendface/LikesController.java [hunk 35]
```diff
@@ -0,0 +1,26 @@
...
+import jakarta.validation.Valid;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+@CrossOrigin(origins = "http://localhost:4200")
+@RestController
+@RequestMapping("likes")
+public class LikesController {
+    private final LikesService likesService;
+    public LikesController(LikesService likesService, HandleValidationExceptions handleValidationExceptions) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LikesController.java]: Do you ever use the exception here

File: friendface-web/src/main/java/com/example/friendface/LikesController.java [hunk 36]
```diff
@@ -0,0 +1,26 @@
...
+        this.likesService = likesService;
+    }
+    @PostMapping
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LikesController.java:31]: Would this be better as a PUT mapping? Considering you are updating something here

File: friendface-web/src/main/java/com/example/friendface/LikesService.java [hunk 37]
```diff
@@ -0,0 +1,37 @@
+package com.example.friendface;
+
+
+import org.springframework.stereotype.Service;
+
+@Service
+public class LikesService {
+    private final LikesRepository likesRepository;
+
+    public LikesService(LikesRepository likesRepository, UserRepository userRepository, PostRepository postRepository) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LikesService.java]: Unused parameters.

File: friendface-web/src/main/java/com/example/friendface/LikesService.java [hunk 38]
```diff
@@ -0,0 +1,37 @@
...
+        this.likesRepository = likesRepository;
+    }
+
+
+    public boolean checkIfLiked(String username, int postId) {
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LikesService.java]: This is very similar to updateLike - you could probably simplify this into one and just ruse it rather than defining more functions.

File: friendface-web/src/main/java/com/example/friendface/LoginService.java [hunk 39]
```diff
@@ -0,0 +1,17 @@
...
+@Service
+public class LoginService {
+    private final UserRepository userRepository;
+
+    public LoginService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    public User findByUsername(String username) {
+        return userRepository.findById(username).orElse(null);
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/LoginService.java]: avoid retuning null. Consider optionals here.

File: friendface-web/src/main/java/com/example/friendface/Post.java [hunk 40]
```diff
@@ -0,0 +1,37 @@
...
+@Entity
+@Table(name="posts")
+@Getter @Setter
+public class Post {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column
+    private String contents;
```
Comment (MMak1599) [inline friendface-web/src/main/java/com/example/friendface/Post.java:22]: validation to avoid saving empty posts, etc. ?

File: friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java [hunk 41]
```diff
@@ -0,0 +1,98 @@
...
+    @MockBean
+    LikesService likesService;
+    @MockBean
+    PostService postService;
+
+
+    @Autowired
+    MockMvc mockMvc;
+    @MockBean
+    private JwtUtil jwtUtil;
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java:36]: Are you using this anywhere? Same for the two below.

File: friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java [hunk 42]
```diff
@@ -0,0 +1,98 @@
...
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@ExtendWith(SpringExtension.class)
+@WebMvcTest(LikesController.class)
+public class LikesControllerTest {
+    @MockBean
+    LikesService likesService;
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java:28]: Is this used?

File: friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java [hunk 43]
```diff
@@ -0,0 +1,98 @@
...
+    private JwtUtil jwtUtil;
+    @MockBean
+    private LoginService loginService;
+    @MockBean
+    private UserService userService;
+
+
+    @Test
+    @WithMockUser
+    void countLikes() throws Exception {
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java:45]: Test name is too generic - test names should reflect scenario + expectation. This is fine for a smaller project, but client projects will have so many tests so having generic naming will lead to mix-up.

File: friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java [hunk 44]
```diff
@@ -0,0 +1,98 @@
...
+                .param("postId", String.valueOf(postId)))
+                .andExpect(status().isOk())
+                .andExpect(content().string(String.valueOf(likes)));
+        verify(postService).countLikes(postId);
+    }
+
+    @Test
+    @WithMockUser(roles = "ADMIN")
+    void checkLike_userHasLiked() throws Exception {
+        LikeDto dto = new LikeDto();
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java:61]: If you are going to build a LikeDto in multiple tests you should consider extracting it out into a helper function, then just calling that helper to construct the DTO.

File: friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java:1]: In these tests you are declaring: 
`ResultActions result = ....,`, then never actually using result afterwards. If you don't need `result` then you can just inline the call.

Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesControllerTest.java:1]: In this file you are only testing happy paths. You should also be adding tests for things such as validation.

File: friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java [hunk 45]
```diff
@@ -0,0 +1,78 @@
...
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class LikesServiceTest {
+    
+    @Mock
+    private LikesRepository likesRepository;
+    
+    @InjectMocks
+    LikesService likesService;
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java:19]: Add a modifier.

File: friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java:1]: DTO setup is repetitive - if you are using it in multiple tests, extract it out into a helper function to make it more readable and concise.

File: friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java [hunk 46]
```diff
@@ -0,0 +1,78 @@
...
+
+        Likes like = new Likes();
+        like.setUsername("Dan");
+        like.setPostId(22L);
+        like.setId(1);
+        
+        when(likesRepository.findByUsernameAndPostId(dto.getUsername(), dto.getPostId())).thenReturn(like);
+        boolean result = likesService.hasUserLiked(dto);
+
+        assertEquals(true, result);
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java:35]: This is valid, but prefer to use: 
`assertTrue(result)`. It gives better failure messages and semantic readability.

File: friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java:1]: Consider adding tests for more than just the happy paths. Tests should also be used for negatives and edge cases.

File: friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java [hunk 47]
```diff
@@ -0,0 +1,78 @@
...
+        LikeDto dto = new LikeDto();
+        dto.setUsername("Dan");
+        dto.setPostId(22L);
+
+        Likes like = new Likes();
+        like.setUsername("Dan");
+        like.setPostId(22L);
+        like.setId(1);
+        
+        when(likesRepository.findByUsernameAndPostId(dto.getUsername(), dto.getPostId())).thenReturn(like);
```
Comment (MMak1599) [inline friendface-web/src/test/java/com/example/friendface/LikesServiceTest.java:32]: You could add interaction verification here - there is no way of seeing if `likesRepository.findByUsernameAndPostId` is called.
